README.md

# REDI <img src="man/figures/logo_redi.png" align="right" width="120" />

<!-- badges: start -->
<!-- badges: end -->

The *REDI* package implements the Robust Exponential Decreasing Index
(REDI). It represents a measure of cumulated workload, is robust to
missing data and provides control of the decreasing influence of the
workload over time.

*REDI* provides various functions to format data, compute REDI and
visualise results in a simple and convenient way.

Issa Moussa, Arthur Leroy et al. (2019): Robust Exponential Decreasing
Index (REDI): adaptive and robust method for computing cumulated
workload. *BMJ Open Sport & Exercise Medicine*,
<https://bmjopensem.bmj.com/content/bmjosem/5/1/e000573.full.pdf>.

## Installation

You can install the development version of REDI from
[GitHub](https://github.com/) with:

``` r
#install.packages("devtools")
devtools::install_github("Grenouil/REDI")
```

## REDI in a nutshell

Here is a basic example on how to simulate a dataset with the adequate
format, compute REDI values over time and display results.

``` r
library(REDI)
set.seed(42)
## Generate a synthetic dataset with the correct format
data = simu_db()

data
#> # A tibble: 366 × 2
#>    Input      Output
#>    <date>      <dbl>
#>  1 2022-01-01   54.3
#>  2 2022-01-02   NA  
#>  3 2022-01-03   51.1
#>  4 2022-01-04   NA  
#>  5 2022-01-05   NA  
#>  6 2022-01-06   49.7
#>  7 2022-01-07   54.8
#>  8 2022-01-08   NA  
#>  9 2022-01-09   NA  
#> 10 2022-01-10   NA  
#> # … with 356 more rows

## Compute REDI over all observations for different Lambda values and display results
res = redi(data)
```

<img src="man/figures/README_REDI_in_a_nutshell-1.png" width="80%" />

``` r

res
#> # A tibble: 1,098 × 4
#>    Input      Output  REDI Lambda
#>    <date>      <dbl> <dbl>  <dbl>
#>  1 2023-01-01   NA    49.2   0.05
#>  2 2022-12-31   49.5  49.2   0.05
#>  3 2022-12-30   NA    49.2   0.05
#>  4 2022-12-29   NA    49.2   0.05
#>  5 2022-12-28   NA    49.2   0.05
#>  6 2022-12-27   47.3  49.2   0.05
#>  7 2022-12-26   45.0  49.3   0.05
#>  8 2022-12-25   49.1  49.7   0.05
#>  9 2022-12-24   52.6  49.7   0.05
#> 10 2022-12-23   NA    49.5   0.05
#> # … with 1,088 more rows
```

## Detailed workflow

For an advanced use of the package, here is a quick step-by-step guide.

### Data simulation

In order to test different features of the package, the `simu_db()`
function is provided as a handy way to generate synthetic data with the
correct format for subsequent REDI computations.

``` r
set.seed(42)
## Generate a synthetic dataset, containing dates (Inputs) from '2022-01-01' to '2023-01-01' and observations (Outputs) following a Gaussian distribution (mean = 50, var = 10), with 50% of missing values.
simu_data <- simu_db(start_date = '2022-01-01',
    end_date = '2023-01-01',
    by = 'day',
    output_distrib = 'Gaussian',
    ratio_missing = 0.5,
    mean = 50,
    var = 10)

simu_data
#> # A tibble: 366 × 2
#>    Input      Output
#>    <date>      <dbl>
#>  1 2022-01-01   54.3
#>  2 2022-01-02   NA  
#>  3 2022-01-03   51.1
#>  4 2022-01-04   NA  
#>  5 2022-01-05   NA  
#>  6 2022-01-06   49.7
#>  7 2022-01-07   54.8
#>  8 2022-01-08   NA  
#>  9 2022-01-09   NA  
#> 10 2022-01-10   NA  
#> # … with 356 more rows
```

As displayed above, any dataset processed in REDI should provide 2
columns: one corresponding to `Input` values (*e.g.* time) and another
to `Output` values (*e.g.* workload).

### Convert a dataset to the correct format with `format_data()`

However, a real-life dataset will generally not have the correct format
to compute REDI directly. Therefore, the `format_data()` function is
designed to help with this process by identifying the columns
corresponding to the Input (e.g the date) and the Output (e.g. the
workload) variables. The Input column should be defined with a correct
`Date` type. The function will automatically identify missing values
between each observations, considering the `by` argument as the time
increment (for instance with ‘day’, the default, each day between two
observed dates is considered missing). Finally, Output values for
duplicated Input values can be summarised according to the
`summarise_duplicate` argument.

``` r

## Create a dummy real-life dataset 
raw_db <- data.frame(
  'Var1' = 1:100, 
  'Var2' = rnorm(n = 100, mean = 50, sd = 10), 
  'Var3' = c(
    as.Date("2022/1/1"), 
    seq(from = as.Date("2022/1/3"), by = "day", length.out = 99)
    )
  )

head(raw_db)
#>   Var1     Var2       Var3
#> 1    1 58.84844 2022-01-01
#> 2    2 51.63431 2022-01-03
#> 3    3 41.39630 2022-01-04
#> 4    4 59.74555 2022-01-05
#> 5    5 43.81340 2022-01-06
#> 6    6 44.72518 2022-01-07

## Convert the dataset to the correct format (adding the missing data point on 2022-01-02) 
db <- format_data(
    data = raw_db,
    input = 'Var3',
    output = 'Var2',
    by = 'day',
    format = '%Y%m%d'
    )

db
#> # A tibble: 101 × 2
#>    Input      Output
#>    <date>      <dbl>
#>  1 2022-01-01   58.8
#>  2 2022-01-02   NA  
#>  3 2022-01-03   51.6
#>  4 2022-01-04   41.4
#>  5 2022-01-05   59.7
#>  6 2022-01-06   43.8
#>  7 2022-01-07   44.7
#>  8 2022-01-08   32.4
#>  9 2022-01-09   45.0
#> 10 2022-01-10   46.8
#> # … with 91 more rows
```

### Computation of REDI with `compute_redi()`

To compute a single REDI value, simply use the `compute_redi()`
function. It will correspond to the REDI value for the most recent
observed Input in the dataset, using all data from the past. Feel free
to adapt the $\lambda$ coefficient, controlling the exponential decay of
weights over time, depending on the context.

``` r

## Compute REDI at the current date (2022-04-11 in this example)
compute_redi(db, coef = 0.05)
#> [1] 49.85942

max(db$Input)
#> [1] "2022-04-11"
```

### Computation of multiple REDI values over time with `loop_redi()`

To sequentially compute REDI for all `Input` values in the dataset with
speed-up vectorised operations, one can use the `loop_redi()` function.

``` r
## Apply loop_redi() to compute REDI for all dates in the dataset 
db_redi <- loop_redi(data = db, coef = 0.5)

db_redi
#> # A tibble: 101 × 4
#>    Input      Output  REDI Lambda
#>    <date>      <dbl> <dbl>  <dbl>
#>  1 2022-04-11   50.1  51.4    0.5
#>  2 2022-04-10   63.9  52.3    0.5
#>  3 2022-04-09   40.9  44.9    0.5
#>  4 2022-04-08   48.3  47.4    0.5
#>  5 2022-04-07   48.7  46.8    0.5
#>  6 2022-04-06   37.9  45.6    0.5
#>  7 2022-04-05   54.4  50.6    0.5
#>  8 2022-04-04   36.9  48.2    0.5
#>  9 2022-04-03   55.0  55.5    0.5
#> 10 2022-04-02   63.0  55.8    0.5
#> # … with 91 more rows
```

### Display results with `plot_redi()`

The `plot_redi()` function is proposed to display the results, and
provides several options to personalise the graphs.

``` r
## Display results as time series of REDI values
plot_redi(redi = db_redi,
          x_axis = 'Date',
          y_axis = 'Workload',
          plot_data = TRUE)
#> Warning: Removed 1 rows containing missing values (`geom_point()`).
```

<img src="man/figures/README_plot_REDI-1.png" width="80%" />

One can customise graphs by:

- changing the name of the axis by replacing the default values of
  `x_axis` and `y_axis`;
- masking the original data in black points by setting `plot_data` to
  `FALSE`.

### Full workflow with the wrapper function `redi()`

As presented in the nutshell example, all the previous steps
(formatting, computations, plotting) are wrapped into the `redi()`
function. In addition to the arguments of the previous functions, it is
also possible to provide a vector to the `coef` argument to display
results for different $\lambda$ values.

``` r
## Apply redi() on db and provide a vector of coefficients.
db_full_redi <- redi(data = db, coef = c(0.1, 0.2, 0.3), plot = TRUE)
```

<img src="man/figures/README_wrapper_REDI-1.png" width="80%" />

``` r

db_full_redi
#> # A tibble: 303 × 4
#>    Input      Output  REDI Lambda
#>    <date>      <dbl> <dbl>  <dbl>
#>  1 2022-04-11   50.1  50.2    0.1
#>  2 2022-04-10   63.9  50.2    0.1
#>  3 2022-04-09   40.9  48.7    0.1
#>  4 2022-04-08   48.3  49.5    0.1
#>  5 2022-04-07   48.7  49.7    0.1
#>  6 2022-04-06   37.9  49.8    0.1
#>  7 2022-04-05   54.4  51.0    0.1
#>  8 2022-04-04   36.9  50.7    0.1
#>  9 2022-04-03   55.0  52.1    0.1
#> 10 2022-04-02   63.0  51.8    0.1
#> # … with 293 more rows
```


NEWS.md
## REDI 1.0.0
Initial release

R/wrapper.R
#' Compute REDI for all observed and missing input values in a dataset
#'
#' Wrapper function that converts the dataset to the adequate format, compute
#' values of REDI for each `Input` values, display a generic plot of the results
#' and return a tibble containing both data and corresponding REDI values.
#'
#' @param data A tibble or a data frame, containing an `Input` column that is
#'    used as reference for the observations (e.g. time for
#'    longitudinal data), and an `Output` column specifying the observed
#'    values of interest (the workload).
#' @param coef A number or vector, containing the values of the lambda
#'    coefficient used in the REDI computations, controlling the decay
#'    of the exponential weights. Default is c(0.05, 0.1, 0.5).
#' @param input A character or a number, indicating the name or the
#'    index of the `Input` column (time).
#' @param output A character or a number, indicating the name or the
#'    index of the `Output` column (workload).
#' @param plot A boolean, indicating whether results should be displayed.
#'    is TRUE.
#' @param by A number or a character string, indicating the reference time
#'    period between two observations. Possible values are 'day', 'week',
#'    'month', 'year', or any arbitrary number. See documentation of the 'seq()'
#'    for additional information if necessary. Default is 'day'.
#' @param format A character string, indicating the date format of the input.
#'    Please read `lubridate::as_date()`. Default is '%Y%m%d'.
#' @param summary_duplicate A function, used to summarise Output values for
#'    duplicated Input values. Default is mean.
#'
#' @return A tibble containing 4 columns : `Input` (without duplicates),
#'        `Output`, `Lambda` and `REDI`, which corresponds to the vector
#'         returned by the `loop_REDI()` function.
#' @export
#'
#' @examples
#' data <- simu_db()
#' redi <- redi(data)
redi <- function(
    data,
    coef = c(0.05, 0.1, 0.5),
    input = 1,
    output = 2,
    plot = TRUE,
    by = 'day',
    format = '%Y%m%d',
    summary_duplicate = mean
    ){

  ## Format the dataset
  data <- format_data(
    data = data,
    input = input,
    output = output,
    by = by,
    format = format,
    summary_duplicate = summary_duplicate)

  ## Compute loop_REDI for all `Lambda` values
  res <- c()
  for (i in coef){
    res <- res %>%
      dplyr::bind_rows(
        loop_redi(data, coef = i) %>% dplyr::mutate('Lambda' = i))
  }

  ## Display a generic plot of the results
  if(plot){
    plot_redi(res) %>% print() %>% suppressWarnings()
  }

  return(res)
}


R/REDI-package.R
#' REDI: Robust Exponential Decreasing Index
#'
#' Implementation of the Robust Exponential Decreasing Index (REDI).
#' The REDI represents a measure of cumulated workload, robust to missing data,
#' providing control of the decreasing influence of workload over time.
#' Various functions are provided to format data, compute REDI, and
#' visualise results in a simple and convenient way.
#'
#' @section Details:
#' For a quick introduction to \pkg{REDI}, please refer to the README at
#' \url{https://github.com/ArthurLeroy/REDI}
#'
#' @section Author(s):
#' Arthur Leroy, Alexia Grenouillat \cr
#' Maintainer: Arthur Leroy - \email{arthur.leroy.pro@@gmail.com}
#'
#' @section References:
#' Issa Moussa, Arthur Leroy et al., BMJ SEM (2019)
#' \url{https://bmjopensem.bmj.com/content/bmjosem/5/1/e000573.full.pdf}
#'
#' @section Examples:
#'
#' ### Simulate a dataset, compute REDI and display results \cr
#' set.seed(42) \cr
#' data <- simu_db(nb_inputs) \cr
#' redi <- redi(data) \cr
#' plot_redi(redi) \cr
#'
#' @docType package
#' @name REDI
NULL
#> NULL


R/plot-functions.R
#' Display the evolution of REDI over time and data points.
#'
#' @param redi A tibble or data frame containing 4 mandatory columns : `Input`,
#'    `Output`, `REDI` and `Lambda`. One can use the `redi()`
#'    function to get this tibble under the right format.
#' @param x_axis A character string, label of the x-axis. Default is 'Input'.
#' @param y_axis A character string, label of the y-axis. Default is 'Output'.
#' @param plot_data A boolean, indicating whether original data should be
#'    displayed. Default is TRUE.
#' @param alpha A number, between 0 and 1, controlling the transparency of
#'                    data points. Default is 0.2.
#' @param size A number, controlling the size of the data points. Default is 1.
#'
#' @return Graph of the evolution of REDI over time, possibly for different
#'    values of Lambda, along with the original data points.
#' @export
#'
#' @examples
#' TRUE
plot_redi <- function(
    redi,
    plot_data = TRUE,
    x_axis = 'Input',
    y_axis = 'Output',
    alpha = 0.2,
    size = 1){

  ## Transform `Lambda` into factor for colouring purposes
  db_plot <- redi %>% dplyr::mutate('Lambda' = as.factor(.data$Lambda))

  ## Plot REDI curves for all values over time of `Lambda`
  gg <- ggplot2::ggplot(db_plot) +
    ggplot2::geom_line(
      ggplot2::aes(
        x = .data$Input,
        y = .data$REDI,
        col = .data$Lambda)
      ) +
    ggplot2::labs(
      x = x_axis,
      y = y_axis,
      title = 'Evolution of REDI over time',
      col = expression(lambda)) +
    ggplot2::theme_classic()

  ## Add raw data points if requested
  if(plot_data){
    gg <- gg +
      ggplot2::geom_point(
        ggplot2::aes(x = .data$Input, y = .data$Output),
        alpha = alpha,
        size = size
      )
  }

  return(gg)
}


R/computing-functions.R
#' Compute REDI for a specific input
#'
#' @param data A tibble or data frame, containing an `Input` column (with the
#'    Date format) and an `Output` column. A simple vector of workload values,
#'    pre-sorted by chronological order, can also be provided.
#' @param coef A number corresponding to the lambda coefficient, controlling the
#'    decay of the exponential weights.
#'
#' @return A number, corresponding to the REDI value at the last `Input` time,
#'    computed over the whole period.
#' @export
#'
#' @examples
#' data <- simu_db()
#' compute_redi(data = data, coef = 0.1)
compute_redi <- function(data, coef = 0.1){
  load <- 0
  norm <- 0

  ## If a vector is provided, create a dummy tibble
  if(data %>% is.vector()){
    data <- tibble::tibble(
      'Input' = 1:length(data),
      'Output' = data)
  }

  ## Sort according to the Input values
  data <- data %>% dplyr::arrange(.data$Input)

  if(any(base::duplicated(data$Input)) == TRUE){
    stop('Some input values are duplicated. Please use format_data().')
  }

  ## Count the number of time points
  nb_inputs <- nrow(data)

  ## Iterate on all inputs starting from 0 to have first coef = 1
  for (i in 0:(nb_inputs - 1))
  {
    ## We compute from present to past, so we are moving backwards
    ## Skip the summation if the value is missing
    if( !is.na(data$Output[nb_inputs - i]) )
    {
      ## Compute the sum of daily loads weighted by the appropriate coefficient
      load <- load + exp(- coef * i) * data$Output[nb_inputs - i]
      ## Update the normalisation constant
      norm <- norm + exp(- coef * i)
    }
  }
  return(load/norm)
}

#' Compute the evolution of REDI over successive inputs
#'
#' @param data A tibble or data frame, containing an `Input` column (with the
#'    Date format) and an `Output` column. A simple vector of workload values,
#'    pre-sorted by chronological order can also be provided.
#' @param coef A number corresponding to the lambda coefficient, controlling the
#'    decay of the exponential weights. Default is 0.1.
#'
#' @return A tibble similar to `data`, containing an additional \code{REDI}
#'     column computed over the successive input values.
#' @export
#'
#' @examples
#' data <- simu_db()
#' loop_redi(data = data, coef = 0.1)
loop_redi <- function(data, coef = 0.1)
{
  ## If a vector is provided, create a dummy tibble
  if(data %>% is.vector()){
    data <- tibble::tibble(
      'Input' = length(data):1,
      'Output' = data)
  }

  ## Sort according to the Input values.
  data <- data %>% dplyr::arrange(dplyr::desc(.data$Input))

  if(any(base::duplicated(data$Input)) == TRUE){
    stop('Some input values are duplicated. Please use format_data().')
  }

  ## Throw an error if the dataset starts with a missing point.
  if(data %>%
     dplyr::filter(.data$Input == min(.data$Input)) %>%
     dplyr::pull(.data$Output) %>%
     is.na()){
    stop("The first Input corresponds to a missing Output value. ",
         "Please provide a first observed value to start computing REDI.")
  }

  ## Compute REDI for all Input values with speed-up vectorised operations
  list_redi <- data %>%
    tidyr::uncount(weights = 1:nrow(data), .id = 'Repet') %>% ## Duplicate DBs
    dplyr::mutate('Coef' = exp(- coef * (.data$Repet - 1))) %>% ## Compute coefs
    dplyr::group_by(.data$Input) %>%
    dplyr::mutate('Sub_db' = dplyr::n() - .data$Repet) %>% ## Identify sub-DBs
    tidyr::drop_na() %>% ## Ignore the missing `Output` values from computations
    dplyr::group_by(.data$Sub_db) %>%
    dplyr::summarise( ## Compute REDI for each sub-DB
      'REDI' = sum( .data$Output * .data$Coef / sum(.data$Coef))) %>%
    dplyr::pull(.data$REDI) ## Extract REDI values

  ## Add REDI values to the dataset and return results
    data %>%
    dplyr::mutate('REDI' = list_redi, 'Lambda' = coef) %>%
    return()
}


R/formatting-functions.R
#' Format the dataset to the syntax of REDI functions
#'
#' @param data A tibble or data frame containing one column indicating time and
#'    another indicating the quantity for which we want to compute REDI.
#' @param input A character or a number, indicating whether the name or the
#'    index of the input column (time).
#' @param output A character or a number, indicating whether the name or the
#'    index of the output column (workload).
#' @param by A number or a character string, indicating the reference time
#'    period between two observations. Possible values are 'day', 'week',
#'    'month', 'year', or any arbitrary number. See documentation of the 'seq()'
#'    for additional information if necessary. Default is 'day'.
#' @param format A character string, indicating the date format of the input.
#'    Please read `lubridate::as_date()`. Default is '%Y%m%d'.
#' @param summary_duplicate A function, used to summarise Output values for
#'    duplicated Input values. Default is mean.
#'
#' @return A tibble with Input and Output columns and explicit missing values
#'    between observations.
#' @export
#'
#' @examples
#' TRUE
format_data <- function(
    data,
    input = 1,
    output = 2,
    by = 'day',
    format = '%Y%m%d',
    summary_duplicate = mean){

  ## Recover the adequate Input and Output column in the dataset
  db <- tibble::tibble(
    'Input' = data[[input]] %>% lubridate::as_date(format = format),
    'Output' = data[[output]]
  )

  ## Test whether there are `Input` values duplicated in the data
  if(any(base::duplicated(db$Input)) == TRUE){
    warning('At least one `Input` value is duplicated. For each duplicated',
            '`Input`, the mean value of the corresponding`Output`is computed',
            ' and only one occurence of `Input` is retained. One can replace',
            ' the mean by another basic function by changing the default value',
            'of `summary_duplicate`.')

    ## Summarise duplicated values and add missing Inputs between observed data
    db <- db %>%
      dplyr::group_by(.data$Input) %>%
      dplyr::summarise('Output' = summary_duplicate(.data$Output)) %>%
      dplyr::distinct(.data$Input, .keep_all = TRUE)
  }

    db %>%
      dplyr::full_join(tibble::tibble(
        'Input' = seq(min(db$Input), max(db$Input), by = by)
      ), by = 'Input') %>%
      dplyr::arrange(.data$Input) %>%
    return()
}


R/simulate-data.R
#' Generate a synthetic dataset tailored for REDI computations
#'
#' Simulate a complete training dataset, which may be representative of various
#' applications. Several flexible arguments allow adjustment of the range of
#' observed days, the distribution and the mean of `Output` values, as well as
#' the ratio of missing data.
#'
#' @param start_date A date, indicating the starting time of observations.
#'    Default is '2022-01-01'.
#' @param end_date A date, indicating the ending time of observations. Default
#'    is '2023-01-01'.
#' @param by A number or a character string, indicating the reference time
#'    time period between two observations. Possible values are 'day', 'week',
#'    'month', 'year', or any arbitrary number. See documentation of the 'seq()'
#'    for additional information if necessary. Default is 'day'.
#' @param output_distrib A character string, indicating the distribution of
#'    `Output` values. Possible values: 'Gaussian' (default), 'Uniform'.
#' @param ratio_missing A number, between 0 and 1, indicating the ratio of
#'    missing values in the dataset. Default is 0.5.
#' @param mean A number, indicating the mean value of the Gaussian distribution.
#'    Default is 50.
#' @param var A number, indicating the variance of the Gaussian distribution.
#'    Default is 10.
#' @param range_unif A vector, indicating the range of values for the Uniform
#'    distribution. Default is c(0,100).
#'
#' @return  A full dataset of synthetic data.
#' @export
#'
#' @importFrom rlang .data
#'
#' @examples
#'
#' ## Generate a dataset with Gaussian measurements
#' data = simu_db(output_distrib = 'Gaussian')
#'
#' ## Generate a dataset with Uniform measurements and 30% of missing data.
#' data = simu_db(output_distrib = 'Uniform', ratio_missing = 0.3)
#'
simu_db <- function(
    start_date = '2022-01-01',
    end_date = '2023-01-01',
    by = 'day',
    output_distrib = 'Gaussian',
    ratio_missing = 0.5,
    mean = 50,
    var = 10,
    range_unif = c(0, 100)
){
  ## Define the range of observed inputs
  input <- seq(
      lubridate::as_date(start_date),
      lubridate::as_date(end_date),
      by = 'day')

  size <- length(input)

  ## Simulate observed output values
  if(output_distrib == 'Gaussian'){
    output = stats::rnorm(size, mean, sqrt(var))
  } else if (output_distrib == 'Uniform'){
    output = stats::runif(size, range_unif[[1]], range_unif[[2]])
  } else {
    stop("The 'output_distrib' argument should be 'Gaussian' or 'Uniform'.")
  }

  ## Draw random locations of missing data
  missing_index <- sample(2:size, ratio_missing * size)

  ## Replace values by NA to simulate missing data
  output[missing_index] <- NA

  ## Create and return the dataset
  tibble::tibble(
    'Input' = input,
    'Output' = output) %>%
   return()
}
